frameworks.qmd

```diff
@@ -1,119 +1,1793 @@
+proportional to computing the function itself. Intricacies of automatic
+differentiation are not dealt with by end users now, but resources to
+learn more can be found widely, such as from
+[[here]{.underline}](https://www.cs.toronto.edu/~rgrosse/courses/csc321_2018/slides/lec10.pdf).
+Automatic differentiation and differentiable programming today is
+ubiquitous and is done efficiently and automatically by modern machine
+learning frameworks.
+
+### Hardware Acceleration
+
+The trend to continuously train and deploy larger machine learning
+models has essentially made hardware acceleration support a necessity
+for machine learning platforms. Deep layers of neural networks require
+many matrix multiplications, which attracts hardware that can compute
+matrix operations fast and in parallel. In this landscape, two types of
+hardware architectures, the [[GPU and
+TPU]{.underline}](https://cloud.google.com/tpu/docs/intro-to-tpu), have
+emerged as leading choices for training machine learning models.
+
+The use of hardware accelerators began with
+[[AlexNet]{.underline}](https://proceedings.neurips.cc/paper_files/paper/2012/file/c399862d3b9d6b76c8436e924a68c45b-Paper.pdf),
+which paved the way for future works to utilize GPUs as hardware
+accelerators for training computer vision models. GPUs, or Graphics
+Processing Units,
+
+excel in handling a large number of computations at once, making them
+ideal for the matrix operations that are central to neural network
+training. Their architecture, designed for rendering graphics, turns out
+to be perfect for the kind of mathematical operations required in
+machine learning. While they are very useful for machine learning tasks
+and have been implemented in many hardware platforms, GPU's are still
+general purpose in that they can be used for other applications.
+
+On the other hand, [[Tensor Processing
+Units]{.underline}](https://cloud.google.com/tpu/docs/intro-to-tpu)
+(TPU) are hardware units designed specifically for neural networks. They
+focus on the multiply and accumulate (MAC) operation, and their hardware
+essentially consists of a large hardware matrix that contains elements
+efficiently computing the MAC operation. This, called the [[systolic
+array
+architecture]{.underline}](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=1653825),
+was pioneered in 1979 by HT Kung and Charles E. Leiserson, but has
+proven to be a useful structure to efficiently compute matrix products
+and other operations within neural networks (such as convolutions).
+
+While TPU's can drastically reduce training times, it also has
+disadvantages. For example, many operations within the machine learning
+frameworks (primarily TensorFlow here since the TPU directly integrates
+with it) are not supported with the TPU's. It also cannot support custom
+custom operations from the machine learning frameworks, and the network
+design must closely align to the hardware capabilities.
+
```

Comment:
It would be nice to add a more clear comparison between GPUs and TPUs here. What are the tradeoffs of using GPUs vs. TPUs and how does this change depending on use case?